Eastern-Research-Group/expert-query
# Feature/232 limit queries

## Related Issues:
* [EQ-232](https://jira.epa.gov/browse/EQ-232)

## Main Changes:
* Added a configurable limit to query counts and selects.
* Moved the "max query size" check entirely to the server to avoid setting the configuration in both the client and server.
* Added the query size check to the `executeQuery` function. This can be bypassed, i.e. if `executeQueryCountOnly` was already called, but the subsequent query will still be limited by the MAX_QUERY_SIZE.
  * This bypass is really just intended to for use by the client UI, we might not want to include it in the API documentation.

## Steps To Test:
1. Add MAX_QUERY_SIZE to the server's environment variables, and set it to 1,000,000.
2. Start the application.
3. Go to http://localhost:3000/attains/catchmentCorrespondence, and open the Network tab in the browser's developer tools.
4. Click Download with no filters, and note the time it takes to get a response from the `count` endpoint.
5. Stop the server, and change MAX_QUERY_SIZE to 1,000.
6. Repeat steps 3 & 4. The response time should be much shorter.
7. Go to http://localhost:3000/attains/catchmentCorrespondence#assessmentUnitId=9MI, and open the Network tab again.
8. Click the Download button, then Continue. The `count` endpoint should not be requested again (the `count` endpoint does seem to be pinged twice initially, but I believe this has to do with React Strict Mode).
